stetho/src/main/java/com/facebook/stetho/inspector/network/DownloadingAsyncPrettyPrinterFactory.java

```diff
@@ -30,53 +30,77 @@ public AsyncPrettyPrinter getInstance(final String headerName, final String head
       return null;
     }
     String uri = result.getSchemaUri();
-    URL schemaURL = parseURL(uri);
+    final URL schemaURL = parseURL(uri);
     if (schemaURL == null) {
-      return getErrorAsyncPrettyPrinter(headerName, headerValue);
+      String errorMessage = "[Failed to parse header: "
+          + headerName + " : " + headerValue + " ]";
+      return getErrorAsyncPrettyPrinter(errorMessage);
     } else {
-      ExecutorService  executorService = AsyncPrettyPrinterExecutorHolder.getExecutorService();
-      if (executorService == null) {
-        //last peer is unregistered...
-        return null;
-      }
-      final Future<String> response = executorService.submit(new Request(schemaURL));
-      return new AsyncPrettyPrinter() {
-        public void printTo(PrintWriter output, InputStream payload)
-            throws IOException {
-          try {
-            String schema;
+      final String schemaCache;
+      AsyncPrettyPrinterSchemaManager schemaFileManager
```

Comment:
I agree that returning different pretty printers throughout this method is not the best idea. I thought about consolidating all the returned pretty printers into one. The major problem that prevented me from being able to do this is response.get() must be called inside PrettyPrinter's printTo method. By design, only printTo is allowed to block for a specified amount of time. This condition somewhat ties my hands to refactor the codes...